### ISSUE TITLE
Feature request: Reload library while syncing if new books are added.

### ISSUE THREAD

**Issue by PinkMusen** [ckilb/pocketbooksync.koplugin#12]
Most of us add our books with calibre, it's less inserts/removal's of the USB port and allows for use of magnetic attachment as data is not needed.

The drawback here is, that to get the built in library of the pocketbook to update to add the new books to it you need to either reboot the device or connect it to the PC with a USB cable.

I'd like to request adding a reload of the library when new books are added, this would make this addon as complete as can be for pocketbook integration to koreader.

**Comment by liskin** [comment]
This would also be useful when books are removed. :-)

Personally, I use PocketBook's DropBox integration and then I mount the DropBox folder using `rclone mount --vfs-cache-mode=full dropbox: ~/mnt/dropbox` and then connect that folder in Calibre. This is probably because I've been using this long before I switched to KOReader. This has the additional advantage that I can tell the PocketBook to Sync and that also reloads the library.

So I'm guessing I might take a look at how the DropBox app for PocketBook makes the library reload and see if we can replicate that. But in the meantime, can you perhaps try if there's a Sync button on your PocketBook somewhere and whether it reloads the library perhaps?

**Comment by liskin** [comment]
Okay, I've done the research:

We need to do roughly the following:

```c
int task = FindTaskByAppName("scanner.app")
if (task != -1) {
  const char *params = "-scan:<directory to reload>"
  SendRequestTo(task, REQ_OPENBOOK, params, strlen(params) + 1, 0, 2000)
}
```

I haven't written any Lua yet, I've only been doing `openbook -1 /ebrmain/bin/scanner.app -scan:/mnt/ext1/something` via ssh. `OpenBook` runs the above code as well (more or less) but it also makes `scanner.app` the foreground task, and we don't want that.

**Comment by liskin** [comment]
Unfortunately KOReader doesn't seem to have hooks for refreshing the library. There's just [hacks like this](https://github.com/koreader/koreader/blob/41f4b1f2acbd4ea8525cb2cfc1ff355543577cbe/plugins/calibre.koplugin/wireless.lua#L46-L56) to refresh KOReader's file browser if it happens to be open. But we'd need to hook into other places as well, like [when you finish a book and get the option to delete it](https://github.com/koreader/koreader/blob/41f4b1f2acbd4ea8525cb2cfc1ff355543577cbe/frontend/apps/reader/modules/readerstatus.lua#L189-L200). That might need some refactoring in KOReader itself :-/

**Comment by liskin** [comment]
I've implemented this as a menu entry for the time being: https://github.com/ckilb/pocketbooksync.koplugin/compare/main...liskin:pocketbooksync.koplugin:reload-library
If you've been rebooting the device to reload the library, this should be a significant upgrade for you.

Next step is to hook this into KOReader, but that will require some way to hook into all the places where the rescan needs to happen. I don't honestly expect to get to that any time soon, so if you could open an issue at https://github.com/koreader/koreader/issues that would be immensely helpful.

**Comment by ckilb** [comment]
@liskin Is it OK for you, if I add your code into this repository?

Also, I'd be interested if you could quickly tell me how you found out how to do this. Especially, how did you know about scanner.app and it's params?

**Comment by liskin** [comment]
Yeah absolutely, if you think some of the code in my fork is ready, feel free to use it.

How I found this - it was a mix of reverse engineering the Dropbox binary and scanner.app using Ghidra and a lot of trial and error. 🙂

**Comment by liskin** [comment]
Happy to elaborate later but am on my mobile at a party now

**Comment by dmalinovsky** [comment]
> I've implemented this as a menu entry for the time being: [main...liskin:pocketbooksync.koplugin:reload-library](https://github.com/ckilb/pocketbooksync.koplugin/compare/main...liskin:pocketbooksync.koplugin:reload-library) If you've been rebooting the device to reload the library, this should be a significant upgrade for you.

I tried your solution, but it seems a bit drastic for me — after running this function, the library marked all the books in the folder as added today, so sorting by recently added broke. :(

**Comment by liskin** [comment]
> > I've implemented this as a menu entry for the time being: [main...liskin:pocketbooksync.koplugin:reload-library](https://github.com/ckilb/pocketbooksync.koplugin/compare/main...liskin:pocketbooksync.koplugin:reload-library) If you've been rebooting the device to reload the library, this should be a significant upgrade for you.
> 
> I tried your solution, but it seems a bit drastic for me — after running this function, the library marked all the books in the folder as added today, so sorting by recently added broke. :(

All books or all books just added? In my experience, tapping the "Scan Home Dir" menu item does the same thing as PocketBook would do on the next reboot anyway. It discovers all new books and shows them as one small pile (or big, but in my case usually small) on the home screen.

**Comment by Silther** [comment]
Is this already in the plugin?

**Comment by liskin** [comment]
> Is this already in the plugin?

Nope, still only available in https://github.com/ckilb/pocketbooksync.koplugin/compare/main...liskin:pocketbooksync.koplugin:reload-library

It's a hack. We discussed a proper solution here:
* https://github.com/ckilb/pocketbooksync.koplugin/pull/14#issuecomment-4035642523
* https://github.com/ckilb/pocketbooksync.koplugin/pull/14#issuecomment-4041673318

(But given that I personally do not have this use-case, I'm somewhat unlikely to tackle this all by myself any time soon.)

**Comment by PinkMusen** [comment]
> Is this already in the plugin?

I actually don't know, I'm using the same version where I pasted the changes in.

The scan home dir saves my USB cable and port 🤗
Sure a reboot does the same but that takes a lot of time,

**Comment by liskin** [comment]
Actually, since we now have a plugins menu entry anyway for the toggling of hiding finished books, adding an extra entry for this hacky "scan home dir" thing isn't a big deal. If people already use it it's better than nothing I guess? Shall I just merge it to main then?

**Comment by Silther** [comment]
> Shall I just merge it to main then?

Would be pretty nice, I mostly use opds and the new bookfusion plugin to load my books onto the reader and If I would not need to restart to get the books would make it a lot easier.

**Comment by liskin** [comment]
Okay, pushed to main: https://github.com/ckilb/pocketbooksync.koplugin/commit/1eb76b794fb788ee755ea086257625ebf9978cc9
It's in 🛠 → PocketBook Sync and also can be used in Gesture Manager.
